pycbc/waveform/generator.py

```diff
@@ -98,21 +104,103 @@ def generate(self, *args):
-        return self.generator(**self.current_params)
+        return self.generate_from_kwargs(**dict(zip(self.variable_args, args)))
 
     def generate_from_kwargs(self, **kwargs):
         """Generates a waveform from the keyword args. The current params
         are updated with the given kwargs, then the generator is called.
         """
         self.current_params.update(kwargs)
+        return self._generate_from_current()
+
+    def _pregenerate(self):
+        """Allows the current parameters to be manipulated before calling
+        the waveform generator.
+        """
+        pass
+
+    def _postgenerate(self, res):
+        """Allows the waveform returned by the generator function to be
+        manipulated before returning.
+        """
+        return res
+
+    @classmethod
+    def set_pregenerate_func(cls, funcname):
+        cls._pregenerate = getattr(cls, funcname)
+
+    @classmethod
+    def set_postgenerate_func(cls, funcname):
+        cls._postgenerate = getattr(cls, funcname)
+
+    def _gdecorator(generate_func):
+        """A decorator that allows for seemless pre/post manipulation of
+        the waveform generator function.
+        """
+        def dostuff(self):
+            self._pregenerate()
+            res = generate_func(self)
+            return self._postgenerate(res)
+        return dostuff
+
+    @_gdecorator
+    def _generate_from_current(self):
+        """Generates a waveform from the current parameters.
+        """
         return self.generator(**self.current_params)
 
 
-class FDomainCBCGenerator(BaseGenerator):
-    """Uses waveform.get_fd_waveform as a generator function to create frequency-
-    domain CBC waveforms in the radiation frame; i.e., with no detector response
-    function applied. For more details, see BaseGenerator.
+class BaseCBCGenerator(BaseGenerator):
+    """Adds ability to convert from various derived parameters to parameters
+    needed by the waveform generators.
+    """
+    def __init__(self, generator, variable_args=(), **frozen_params):
+        super(BaseCBCGenerator, self).__init__(generator,
+            variable_args=variable_args, **frozen_params)
+        # if m1 and m2 are not parameters, decorate the generator function
+        # to convert the used mass parameters to m1, m2
+        all_args = self.frozen_params.keys() + list(self.variable_args)
+        if 'mass1' not in all_args or 'mass2' not in all_args:
+            # set the decorator to the appropriate converter
+            if 'mchirp' in all_args and 'eta' in all_args:
+                self.set_pregenerate_func('mchirp_eta_to_mass1_mass2')
+            elif 'mtotal' in all_args and 'eta' in all_args:
+                self.set_pregenerate_func('mtotal_eta_to_mass1_mass2')
+            else:
+                raise ValueError("if not specifying mass1, mass2, must either use "
+                    "(mchirp, eta) or (mtotal, eta)")
```

Comment:
__[flake8]__
* [E128] continuation line under-indented for visual indent

<!-- Comment by [SideCI](https://sideci.com) -->